### ISSUE TITLE
data.template_file using count = zero produces error inside module

### ISSUE THREAD

**Issue by Gary-Armstrong** [hashicorp/terraform#16022]
Can generate `data.template_file` using `count = 0` in local .tf files, but not when inside a module.

### Terraform Version
0.9 - 0.10

### Terraform Configuration Files
Working sample from a file `bootstrap.tf` within the dir in which I run `terraform` commands.
```hcl
data "template_file" "bootstrap_testhead" {
  count = "${var.headtest_count}"
  template = "${file("../../config-common/bootstrap.tpl")}"
  vars {
```
When the instance count is zero, no template is generated.

Duplicating this into a module produces the following when count is zero:
```
* module.head.data.template_file.userdata: strconv.ParseInt: parsing "": invalid syntax
```

### Debug Output
Relevant debug output:
```2017/08/31 12:08:42 [TRACE] root: eval: *terraform.EvalWriteState
2017/08/31 12:08:42 [TRACE] [walkRefresh] Exiting eval tree: aws_route53_record.compute-00-public[2]
2017/08/31 12:08:42 [TRACE] dag/walk: walking "root"
2017/08/31 12:08:42 [TRACE] vertex 'root.root': walking
2017/08/31 12:08:42 [TRACE] dag/walk: upstream errored, not walking "provider.aws (close)"
2017/08/31 12:08:42 [TRACE] dag/walk: upstream errored, not walking "root"
2017/08/31 12:08:42 [DEBUG] plugin: waiting for all plugin processes to complete...
```
If this isn't enough, let me know.

### Expected Behavior
What should have happened?
TF should have just skipped over the template file creation, as it does when not run from a module.

### Actual Behavior
What actually happened?
TF decided to error on a empty string.

### Steps to Reproduce
1. Create a resource `aws_instance` that uses a `count` parameter.
2. Define the count in the tfvars file as "0".
3. Create a data resource `data "template_file"` also using same `count` within the same directory.
4. Observe `terraform plan` will happily not create anything.
5. Put that code into another dir; make it a module & push to github.
6. Create a resource `module "instance"`, passing in relevant variables including a zero count.
7. Get the module and plan.
8. Observe TF complain about an empty string.

### Important Factoids
I have asked around the community and as far as I can tell, this should work.  If there is some alternate method, please share.

**Comment by Gary-Armstrong** [comment]
I figured out this much when I changed the count variables in the local tfvars to "1".  The plan succeeds when count is non-zero but it needs to succeed when count is zero.

**Comment by Gary-Armstrong** [comment]
After I wrote this great issue, I went back and set all the counts from "1" to "0" again.  After this, the plan succeeded.  I guess this means there is a workaround but golly gee willikers.  There is some gap that needs to be checked during this case.

**Comment by Gary-Armstrong** [comment]
Some more work on this on my end, and I determined that if I unset the vars in .tfvars and let the `default = "" ` from the variable definition fall through, I get the empty string error.  Which, again, is pretty obvious in hindsight and now I'm struggling to recall if this is the case when doing this without modules.  I think I'd have run into this otherwise.

Anyway, even if this is going to be labeled "user error", can something be done going forward to prevent people like me from wasting a lot of time doing dumb things?  Some validation enhancement?  A documentation tidbit?

**Comment by apparentlymart** [comment]
Hi @Gary-Armstrong! Sorry for this diversion/confusion.

If I'm understanding correctly, it seems like you eventually figured out here that the count was being set to the empty string rather than zero.

I think the _least_ we can do here is improve that awful error message, so that it:
* Identifies that it's the `count` attribute it's talking about
* Doesn't reference a Go function that it happens to be using internally to parse the number
* Possibly even recognizes the empty string as a special case and produces a custom message explaining that it should be set explicitly to zero here.

We're just getting started now on some configuration language changes that include retaining enough information in the config parser that we can produce better error messages than we tend to today. There may be other things we can do here to help with this specific case, but I'd like to see how far we can get with just improving the error message first and then we'll see what, if any, other changes would help make it clearer what's going on here.

It's likely that some initial changes to the error message here will come as part of some broader work, but I'll leave this issue here to remind us to revisit this _specific_ situation and make sure the error message for this is as good as it can be, and to see if there's anything else we can do along with that.

**Comment by Gary-Armstrong** [comment]
Thanks for the response.  It is as you say.  I think I must have previously been relying on some conversion done within the "main" HCL, where empty strings are silently converted to int zero or boolean false as needed.  I can understand that module inputs are passed "as-is".  But I haven't been able to manufacture the time to test this out for realsies.

As for the error message, I can say that Ye Olde ParseInt has shown up a lot for me since 0.6 and I got into the habit of not taking it literally.  Some of this is my mental baggage, but a cleaner message would possibly have helped me realize I need to actually pass in the variable I want. :)

**Comment by mildwonkey** [comment]
Hi!
Terraform 0.12 has both improved type checking and improved error messages which address this issue:

count is `""`:
```
Error: Incorrect value type

  on main.tf line 10, in data "template_file" "my_tmpl":
  10:   count = var.template_count

Invalid expression value: a number is required.
```

count is `null`:
```
Error: Invalid count argument

  on main.tf line 10, in data "template_file" "my_tmpl":
  10:   count = var.template_count

The given "count" argument value is null. An integer is required.
```

I am going to close this as being fixed in 0.12 Thanks!
